File: src/main/java/com/harleylizard/magic_things/common/FouledTreeFeature.java
package com.harleylizard.magic_things.common;

import com.harleylizard.magic_things.common.block.Facing;
import com.harleylizard.magic_things.common.block.FouledLogRootBlock;
import net.minecraft.core.BlockPos;
import net.minecraft.tags.BlockTags;
import net.minecraft.world.level.WorldGenLevel;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.Blocks;
import net.minecraft.world.level.levelgen.feature.Feature;
import net.minecraft.world.level.levelgen.feature.FeaturePlaceContext;
import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration;
import org.jetbrains.annotations.NotNull;

public final class FouledTreeFeature extends Feature<NoneFeatureConfiguration> {

    public FouledTreeFeature() {
        super(NoneFeatureConfiguration.CODEC);
    }

    @Override
    public boolean place(@NotNull FeaturePlaceContext<NoneFeatureConfiguration> context) {
        var random = context.random();
        var height = 4 + random.nextInt(4);

        var level = context.level();
        var pos = context.origin();

        var log = MagicThingsBlocks.FOULED_LOG.defaultBlockState();

        var canPlace = true;
        for (var i = 0; i < height; i++) {
            if (!level.getBlockState(pos.above(i)).is(MagicThingsBlockTags.FOULED_LOG_CAN_REPLACE)) {
                canPlace = false;
                break;
            }
        }

        if (canPlace) {
            for (var i = 0; i < height; i++) {
                level.setBlock(pos.above(i), log, Block.UPDATE_ALL);
            }

            var top = pos.above(height);
            var distance = 0;
            for (var facing : Facing.values()) {
                if (random.nextInt(1) == 0) {
                    var below = facing.relative(top);
                    var i = 0;
                    while (level.getBlockState(below).is(MagicThingsBlockTags.FOULED_GROWTH_CAN_REPLACE)) {
                        below = below.below();
                        if (i++ > 7) {
                            break;
                        }
                    }

                    var floor = below.above();

                    var blockState = level.getBlockState(floor);
                    if (blockState.getFluidState().isEmpty()) {
                        var root = FouledLogRootBlock.set(MagicThingsBlocks.FOULED_LOG_ROOT.defaultBlockState(), facing.getOpposite());

                        if (root.canSurvive(level, floor) && level.setBlock(floor, root, Block.UPDATE_ALL)) {
                            distance = Math.max(distance, top.getY() - floor.getY());
                            placeDirt(level, below);
                        }
                    }
                }
            }

            for (var i = 0; i < distance; i++) {
                level.setBlock(top.below(i + 1), log, Block.UPDATE_ALL);
            }
            return true;
        }

        return false;
    }

    private void placeDirt(WorldGenLevel level, BlockPos blockPos) {
        if (level.getBlockState(blockPos).is(BlockTags.REPLACEABLE)) {
            level.setBlock(blockPos, Blocks.DIRT.defaultBlockState(), Block.UPDATE_ALL);
        }

    }

}

File: src/main/kotlin/com/harleylizard/magic_things/common/terrablender/FouledTreeFeature.kt
package com.harleylizard.magic_things.common.terrablender

import com.harleylizard.magic_things.common.MagicThingsBlocks
import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.tags.BlockTags
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import net.minecraft.world.level.block.state.properties.BlockStateProperties
import net.minecraft.world.level.levelgen.feature.Feature
import net.minecraft.world.level.levelgen.feature.FeaturePlaceContext
import net.minecraft.world.level.levelgen.feature.configurations.NoneFeatureConfiguration

class FouledTreeFeature() : Feature<NoneFeatureConfiguration>(NoneFeatureConfiguration.CODEC) {

    override fun place(context: FeaturePlaceContext<NoneFeatureConfiguration>): Boolean {
        val level = context.level()
        val random = context.random()
        val blockPos = context.origin()

        val height = 4 + random.nextInt(4)

        val top = blockPos.above(height)

        var floor = top.y
        for (direction in Direction.Plane.HORIZONTAL) {
            if (random.nextInt(8) == 0) {
                continue
            }
            var relative = top.relative(direction)

            var i = 0
            while (level.getBlockState(relative.below().also { relative = it }).let { it.`is`(BlockTags.REPLACEABLE) || it.`is`(MagicThingsBlocks.fouledGrowth) }) {
                if (i++ > 8) {
                    break
                }
            }

            val lowest = relative.above()

            if (level.getBlockState(lowest).let { it.fluidState.isEmpty && replaceable(it) }) {
                val root = MagicThingsBlocks.fouledLogRoots.defaultBlockState().setValue(BlockStateProperties.HORIZONTAL_FACING, direction.opposite)

                level.setBlock(lowest, root, Block.UPDATE_ALL)

                floor = Math.min(floor, lowest.y)
            }
        }

        for (i in floor..top.y) {
            val relative =  BlockPos(blockPos.x, i, blockPos.z)

            level.getBlockState(relative).takeIf { replaceable(it) || it.`is`(BlockTags.DIRT) }?.let {
                level.setBlock(relative, MagicThingsBlocks.fouledLog.defaultBlockState(), Block.UPDATE_ALL)
            }

        }

        return true
    }

    fun replaceable(blockState: BlockState) = blockState.`is`(BlockTags.REPLACEABLE_BY_TREES) || blockState.`is`(BlockTags.REPLACEABLE) || blockState.`is`(BlockTags.LUSH_GROUND_REPLACEABLE)

}